File: 프로그래머스/lv2/42839. 소수 찾기/소수 찾기.swift
import Foundation

func solution(_ numbers:String) -> Int {
    var n = numbers.map({ String($0) })
    n.sort(by: >)
    var visited = Array(repeating: false, count: n.count)
    var list = [String]()
    combination(n: n, list: &list, visited: &visited)
    return memo.count
}

var memo = [Int: Bool]()

func combination(n: [String], list: inout [String], visited: inout [Bool]) {
    if let sumList = Int(list.reduce("", +)) {
        if let _ = memo[sumList] {
            return
        }
        
        if sumList.isPrimeNum() {
            memo.updateValue(true, forKey: sumList)
        }
    }
    
    if list.count == n.count {
        return
    }
    
    for i in 0 ..< n.count {
        if !visited[i] {
            visited[i] = true
            list.append(n[i])
            combination(n: n, list: &list, visited: &visited)
            list.removeLast()
            visited[i] = false
        }
    }
}

extension Int {
    // 소수 판별
    func isPrimeNum() -> Bool {
        if self >= 0 {
            if self < 2 {
                return false
            }
            for i in 2 ..< Int(sqrt(Double(self)) + 1) {
                if self % i == 0 {
                    return false
                }
            }
            return true
        }
        return false
    }
}

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.py
from functools import reduce
import math

def solution(numbers):
    answer = 0
    comb = []
    visited = [False for _ in range(len(numbers))]    
    
    def combination(end, arr):
        if arr:
            num = reduce(lambda x, y: x + y, list(arr))
            comb.append(int(num))
        
        for i in range(len(numbers)):
            if not visited[i]:
                visited[i] = True
                arr.append(numbers[i])
                combination(end, arr)
                arr.pop()
                visited[i] = False
                
    combination(len(numbers), [])
    
    for num in set(comb):
        if is_prime_number(num):
            answer += 1
            
    return answer


def is_prime_number(m):
    if m < 2:
        return 0
    elif m == 2:
        return 1
    if m % 2 == 0:
        return 0
    for i in range(3, int(math.sqrt(m)) + 1, 2):
        if m % i == 0:
            return 0
    return 1
        